system/lib/libc/dynlink.c

```diff
@@ -105,7 +107,221 @@ static struct dso* load_library_start(const char* name, int flags) {
   return p;
 }
 
-static void dlopen_js_onsuccess(struct dso* dso, struct async_data* data) {
+// This function is called at the start of all entry points so that the dso
+// list gets initialized on first use.
+static void ensure_init() {
+  if (head) {
+    return;
+  }
+  // Initialize the dso list.  This happens on first run.
+  pthread_rwlock_wrlock(&lock);
+  if (!head) {
+    // Flags are not important since the main module is already loaded.
+    struct dso* p = load_library_start("__main__", RTLD_NOW|RTLD_GLOBAL);
+    assert(p);
+    _dlinit(p);
+    load_library_done(p);
+    assert(head);
+  }
+  pthread_rwlock_unlock(&lock);
+}
+
+#ifdef _REENTRANT
+// These functions are defined in JS.
+
+// Signal the completion of a given promise that was passed into
+// emscripten_proxy_sync_code.
+void _emscripten_promise_resolve(int promise_id, void* value);
+void _emscripten_promise_reject(int promise_id);
+
+// async JS function that runs emscripten_thread_sync_code of each of threads
+// that are running at the time of the call.  Once this is done the callback is
+// called with the given em_proxying_ctx.
+void _emscripten_sync_all_threads(pthread_t calling_thread,
+                                  void (*callback)(em_proxying_ctx*),
+                                  em_proxying_ctx* ctx);
+
+static void sync_next(struct dso* dso, int promise_id);
+
+static void sync_one_onsuccess(struct dso* dso, void* user_data) {
+  int promise_id = (intptr_t)user_data;
+  // Load the next dso in the list
+  thread_local_tail = dso;
+  sync_next(dso->next, promise_id);
+}
+
+static void sync_one_onerror(struct dso* dso, void* user_data) {
+  int promise_id = (intptr_t)user_data;
+  pthread_rwlock_unlock(&lock);
+  _emscripten_promise_reject(promise_id);
```

Comment:
Do we expect to recover from this? My first guess is that this should be a fatal error here, but I'm probably missing something.

Reply:
I guess we should decide to simply return an error from dlopen().. but maybe it would be too hard to try to recover into a sane state.. for now I think are you right.

Reply:
Done